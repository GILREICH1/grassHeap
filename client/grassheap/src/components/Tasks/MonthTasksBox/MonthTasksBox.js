import React, {useState, useEffect } from 'react';
import TaskList from '../TaskList/TaskList'
import { getTasksByMonth } from '../../../services/ServerApiServices';

function MonthsTasksBox({monthNumber, monthName}) {

  const [ tasks, setTasks ] = useState([]);

  useEffect(()=>{
    getTasksByMonth(monthName).then(tasks => setTasks(tasks));
  }, [])

  return (
    <div>
      <h3>{monthName} {monthNumber}</h3>
      <TaskList tasks={tasks} />
    </div>
    )
}

export default MonthsTasksBox;
