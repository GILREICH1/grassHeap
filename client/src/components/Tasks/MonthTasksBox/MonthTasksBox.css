h2::first-letter {
  text-transform: capitalize;
}

.MonthTaskBox {
  height: 40vh;
  overflow-y: scroll;
  overflow-x: hidden;
}

/* Hide scrollbar for Chrome, Safari and Opera */
.MonthTaskBox::-webkit-scrollbar {
  display: none;
}
.MonthTaskBox {
  -ms-overflow-style: none; /* IE and Edge */
  scrollbar-width: none; /* Firefox */
}
