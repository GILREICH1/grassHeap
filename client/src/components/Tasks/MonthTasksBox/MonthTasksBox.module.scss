@import '../../../common/layout.scss';

h2::first-letter {
  text-transform: capitalize;
}
