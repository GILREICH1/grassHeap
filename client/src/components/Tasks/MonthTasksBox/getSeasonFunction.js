export function getSeason(monthNumber) {
  if (monthNumber < 1) monthNumber = 12 + monthNumber;
  let season = '';
  switch (monthNumber) {
    case '12':
    case '1':
    case '2':
      season = '⛄';
      break;
    case '3':
    case '4':
    case '5':
      season = '🌼';
      break;
    case '6':
    case '7':
    case '8':
      season = '🌞';
      break;
    case '9':
    case '10':
    case '11':
      season = '🍂';
      break;
  }
  return season;
}
