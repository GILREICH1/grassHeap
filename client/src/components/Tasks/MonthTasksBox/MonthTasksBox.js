import React, { useState, useEffect } from "react";
import TaskList from "../TaskList/TaskList";
import {
  deleteTask,
  getMyPlants,
  getTasksByMonth,
} from "../../../services/ServerApiServices";
import "./MonthTasksBox.css";
import AddTaskForm from "../AddTaskForm/AddTaskForm";

function getSeason(monthNumber) {
  let season = "";
  switch (monthNumber) {
    case "12":
    case "1":
    case "2":
      season = "⛄";
      break;
    case "3":
    case "4":
    case "5":
      season = "🌼";
      break;
    case "6":
    case "7":
    case "8":
      season = "🌞";
      break;
    case "9":
    case "10":
    case "11":
      season = "🍂";
      break;
  }
  return season;
}

function MonthsTasksBox({ monthNumber, monthName }) {
  const [tasks, setTasks] = useState([]);
  const [seasonIcon, setSeasonIcon] = useState("");

  useEffect(() => {
    getMyPlants().then((myPlants) => {
      getTasksByMonth(monthName).then((tasks) => {
        // filter tasks to those which are relevant to plants saved in myPlants database OR added manually
        const myTasks = tasks.filter((task) =>
          myPlants.some((plant) => plant.name === task.crop || task.userCreated)
        );
        setTasks(myTasks);
      });
    });
  }, [monthNumber, monthName]);

  useEffect(() => {
    setSeasonIcon(getSeason((monthNumber + 1).toString()));
  }, [monthNumber, monthName]);

  function addNewTask(task) {
    setTasks([...tasks, task]);
  }
  function deleteThisTask(_id) {
    setTasks([...tasks].filter((task) => task._id !== _id));
    deleteTask(_id);
  }

  return (
    <div className={`MonthTaskBox__${seasonIcon}`}>
      <h2>
        {monthName} {seasonIcon}
      </h2>
      <TaskList deleteThisTask={deleteThisTask} tasks={tasks} />
      <AddTaskForm addNewTask={addNewTask} month={monthName} />
    </div>
  );
}

export default MonthsTasksBox;
