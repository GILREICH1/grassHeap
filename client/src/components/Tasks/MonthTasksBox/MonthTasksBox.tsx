/* eslint-disable no-unused-vars */
import React, { useState, useEffect, useContext } from 'react';
import TaskList from '../TaskList/TaskList';
import { deleteTask } from '../../../services/ServerApiServices';
import AddTaskForm from '../AddTaskForm/AddTaskForm';
import { getSeason } from './getSeasonFunction';
import { Task } from '../../../common/types';
import { userContxt } from '../../Authentication/UserContext';
import { useAuth0 } from '@auth0/auth0-react';
import { months } from '../../../utils/months';
import Card from '@mui/joy/Card';

const filterUserTasksByMonth = (tasks: Task[], monthName: string) => {
  return tasks.filter(task => task.month === monthName);
};

interface MonthProps {
  monthNumber: number;
  tasks: Task[];
  setTasks: (tasks: Task[]) => void;
}

function MonthTasksBox({
  tasks,
  monthNumber,
  setTasks,
}: MonthProps): JSX.Element {
  const [monthTasks, setMonthTasks] = useState<Task[]>([]);
  const [seasonIcon, setSeasonIcon] = useState('');
  const { isAuthenticated, getAccessTokenSilently } = useAuth0();
  const { user } = useContext(userContxt);
  const monthName = months[monthNumber];

  useEffect(() => {
    setSeasonIcon(getSeason(monthNumber + 1));
  }, [monthNumber]);

  useEffect(() => {
    const monthName = months[monthNumber];
    const filteredTasks = filterUserTasksByMonth(tasks, monthName);
    setMonthTasks(filteredTasks);
  }, [monthNumber, tasks]);

  async function deleteThisTask(_id: string): Promise<void> {
    if (isAuthenticated) {
      const token = await getAccessTokenSilently();
      await deleteTask({ _id, user, token });
    }
    const newTasksList: Task[] = tasks.filter(task => task._id !== _id);
    setTasks(newTasksList);
  }

  return (
    <Card>
      <h2>
        {monthName} {seasonIcon}
      </h2>
      <TaskList deleteThisTask={deleteThisTask} tasks={monthTasks} />
      <AddTaskForm
        addNewTask={(task: Task) => setTasks([...tasks, task])}
        month={monthName}
      />
    </Card>
  );
}

export default MonthTasksBox;
