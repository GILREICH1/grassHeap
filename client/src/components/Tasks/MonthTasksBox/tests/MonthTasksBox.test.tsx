import React from 'react';
import renderer from 'react-test-renderer';
import MonthTasksBox from '../MonthTasksBox';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { getSeason } from '../getSeasonFunction';
import mocks from './mocks';

describe('MonthTasksBox tests', () => {
  let container: Element;
  const monthName = 'January';
  const monthNumber = 1;

  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    if (container) {
      unmountComponentAtNode(container);
      container.remove();
    }
  });

  describe('layout', () => {
    test('snapshot', () => {
      const tree = renderer
        .create(<MonthTasksBox monthName="January" monthNumber={1} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('props', () => {
    test('no tasks', async () => {
      jest.mock('../../../../services/ServerApiServices', () => {
        const getTasksByMonth = jest.fn(() => Promise.resolve([]));

        const deleteTask = jest.fn();
        return { getTasksByMonth, deleteTask };
      });

      act(() => {
        render(
          <MonthTasksBox monthName={monthName} monthNumber={monthNumber} />,
          container,
        );
      });
      const pElement = container.querySelector('p');

      expect(pElement?.textContent).toBe(
        'No tasks for your plants this month!',
      );
    });

    // TODO mock plants && user context
    //   test('tasks', async () => {
    //     jest.mock('../../../../services/ServerApiServices', () => {
    //       const getTasksByMonth = async () => mocks.taskList;

    //       const deleteTask = jest.fn();
    //       return { getTasksByMonth, deleteTask };
    //     });

    //     act(() => {
    //       render(
    //         <MonthTasksBox monthName={monthName} monthNumber={monthNumber} />,
    //         container,
    //       );
    //     });

    //     // expect(container.querySelector("p").textContent).toBe();
    //     expect(container.textContent).toContain(mocks.taskList[0]['task']);
    //   });
  });

  test('renders month season description', () => {
    act(() => {
      render(
        <MonthTasksBox monthName={monthName} monthNumber={monthNumber} />,
        container,
      );
    });

    const seasonEmoji = getSeason(monthNumber);
    const h2Content = container.querySelector('h2');
    expect(h2Content).toBeTruthy();
    expect(h2Content?.textContent).toBe(`${monthName} ${seasonEmoji}`);
  });
});
