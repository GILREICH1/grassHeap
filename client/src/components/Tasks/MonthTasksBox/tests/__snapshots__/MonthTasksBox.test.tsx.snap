// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthTasksBox tests layout snapshot 1`] = `
<div
  className="MonthTaskBox"
>
  <h2>
    January
     
    
  </h2>
  <div
    className="TaskList"
  >
    <div
      className="TaskList_empty"
    >
      <p>
        No tasks for your plants this month!
      </p>
    </div>
  </div>
  <form
    className="submitForm"
    onSubmit={[Function]}
  >
    <h3>
      Add New Task
    </h3>
    <div
      className="form__group"
    >
      <input
        className="form__field"
        onChange={[Function]}
        required={true}
        value=""
      />
      <label
        className="form__label"
      >
        Task
      </label>
    </div>
    <div
      className="form__group"
    >
      <input
        className="form__field"
        list="plant-list"
        onChange={[Function]}
        required={true}
        value=""
      />
      <label
        className="form__label"
      >
        Crop
      </label>
      <datalist
        id="plant-list"
      />
    </div>
    <input
      className="submit"
      type="submit"
      value="Submit"
    />
  </form>
</div>
`;
